### ISSUE TITLE
Exporting often leads to application error

### ISSUE THREAD

**Issue by acz13** [UlyssesInvictus/QuizDB#5]
Searching or randoming then exporting to anything often leads us to the Heroku application error page...

Probably should be doing the exporting from blobs client-side generated from `props.question[tossups/bonuses]`, anyway

**Comment by acz13** [comment]
Looks like the server is just taking a really really long time?

**Comment by UlyssesInvictus** [comment]
I'll take a look. Most likely cause is timeouts caused since exporting questions auto overrides the default limit.

I don't know if I exactly want to set it up to serve exporting as client-side right now: you'd have to generate it from the backend anyways, so it would just increase the time to serve. The UI would be a little weird as well, since it's a hidden critical failure in the app that there's no way to view a **lot** of questions on a page at once--it crashes your browser b/c there's too much on the page--and presumably people would have to search for, and then see, what they want to download.

It's an interesting design question that I'm not really proficient in solving. I will think about it, though.

Let me know if you have any ideas as well.

**Comment by acz13** [comment]
(I haven't had time to look at the Rails backend code yet, so take what I say next with a grain of salt)

- The site reliability is being pretty sketch right now... (does PACE not have hosting capability?)
- Exporting is just completely borked if you get a lot of results (bc limit is disabled)
- Actually the whole site is just giving me errors right now, did something happen?
- I think what needs to be done is some sort of paging system w/ a limit on how many questions you can request at once... Redoing the entire search is kind of iffy just for export (this notably completely fails on random searches). I think it should not be too hard to implement client side export to fix this. I can give you a mockup when I have time if needed (it's super easy though)

**Comment by UlyssesInvictus** [comment]
- PACE isn't actually hosting itself, they're just giving me some money. (Which hasn't come yet either... :( ). So it's still in my interest to use cheaper servers--which I'm doing right now because looking at the logs, it's fine 95% of the time.
- Yeah, the exporting thing is an issue. I'm going to silently limit exports at 500 questions, because that seems like a reasonable limit on what anyone would ever want at once, but I do still think people will want to download more questions than they see on the page. The real issue is just when people do a blank search (like with Random) and then download everything (:looking at you, buddy:) (though it's really my fault for bad design...). In the meantime hopefully people stop requesting everything...
- Looking at the logs, it crashed about when you made this comment because of memory errors. Guessing you had just tried to download everything. The logs look fine after it reboots though.
- Yeah, I'm **definitely** implementing paging. It's easy enough to imagine how do this (server-side, even) for regular question browsing, but I'm still not sure how to do pages when people are actually downloading questions and still want more results than would fit on a page (say, for example, you want every question on Poland). I might just keep the aforementioned 500 question limit and put a comment in the file saying to go directly to the API if they need more. 

Basically, more design considerations.

**Comment by acz13** [comment]
>Guessing you had just tried to download everything

Oops, sorry! I'll set up a local testing instance when I have time (currently just proxying requests w/ webpack server...)

**Comment by UlyssesInvictus** [comment]
Nah, it's on me for even leaving that as a possibility. If you are interested in doing local testing, let me know and I'll redump the DB files to my s3 bucket so you can potentially even grab the current state of the DB.

I really appreciate you looking into the code this closely btw: if you ever feel the need to directly make a PR, just shoot for it!

**Comment by acz13** [comment]
The frontend is static right? That could probably be served with cheap S3 static site hosting to reduce load off Heroku 🤔

**Comment by UlyssesInvictus** [comment]
True, but I made the decision to push my client and my backend in the same repo so that updating/deploying them is easy, with the con that they're the same heroku instance (they're other upsides, like heroku giving me ssl certs, but that's beside the point).

I should probably move them separately, but it's not a big deal currently because the BE workers dwarf the web workers in size; and I use PWA strategies anyways, so most people have the static files cached, with extra pings occasionally to redownload new files.

The real memory save I can do is moving the database to a separate server that I can then speed up separately. It won't solve all the problems, but it'll mean I'm not wasting as much money on Heroku hosting it.
